rescheduler/rescheduler.go

```diff
@@ -77,8 +88,247 @@ func main() {
+			delErr := client.Pods(p.Namespace).Delete(p.Name, kube_api.NewDeleteOptions(10))
+			if delErr != nil {
+				return fmt.Errorf("Failed to delete pod %s: %v", podId(p), delErr)
+			}
+		} else {
+			newPods := append(nodeInfo.Pods(), p)
+			nodeInfo = schedulercache.NewNodeInfo(newPods...)
+			nodeInfo.SetNode(node)
+		}
+	}
+
+	// TODO(piosz): how to reset scheduler backoff?
+
+	return nil
+}
+
+func addTaint(client *kube_client.Client, node *kube_api.Node, value string) error {
+	taints, err := kube_api.GetTaintsFromNodeAnnotations(node.Annotations)
+	if err != nil {
+		return err
+	}
+
+	taint := kube_api.Taint{
+		Key:    criticalAddonsOnlyTaintKey,
+		Value:  value,
+		Effect: kube_api.TaintEffectNoSchedule,
+	}
+	taints = append(taints, taint)
+
+	taintsJson, err := json.Marshal(taints)
+	if err != nil {
+		return err
+	}
+
+	if node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+	node.Annotations[kube_api.TaintsAnnotationKey] = string(taintsJson)
+	_, err = client.Nodes().Update(node)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func findNodeForPod(client *kube_client.Client, predicateChecker *ca_simulator.PredicateChecker, nodes []*kube_api.Node, pod *kube_api.Pod) *kube_api.Node {
+	for _, node := range nodes {
+		// ignore nodes with taints
+		if err := checkTaints(node); err != nil {
+			glog.Warningf("Skipping node %v due to %v", node.Name, err)
+		}
+
+		requiredPods, _, err := groupPods(client, node)
+		if err != nil {
+			glog.Warningf("Skipping node %v due to error: %v", node.Name, err)
+			continue
+		}
+
+		nodeInfo := schedulercache.NewNodeInfo(requiredPods...)
+		nodeInfo.SetNode(node)
+
+		if err := predicateChecker.CheckPredicates(pod, nodeInfo); err == nil {
+			return node
+		}
+	}
+	return nil
+}
+
+func checkTaints(node *kube_api.Node) error {
+	taints, err := kube_api.GetTaintsFromNodeAnnotations(node.Annotations)
+	if err != nil {
+		return fmt.Errorf("error while getting taints: %v", err)
+	}
+	for _, taint := range taints {
+		if taint.Key == criticalAddonsOnly {
+			return fmt.Errorf("CriticalAddonsOnly taint with value: %v", taint.Value)
+		}
+	}
+	return nil
+}
+
+// groupPods divides pods running on <node> into those which can't be deleted and the others
+func groupPods(client *kube_client.Client, node *kube_api.Node) ([]*kube_api.Pod, []*kube_api.Pod, error) {
+	podsOnNode, err := client.Pods(kube_api.NamespaceAll).List(
+		kube_api.ListOptions{FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": node.Name})})
+	if err != nil {
+		return []*kube_api.Pod{}, []*kube_api.Pod{}, err
+	}
+
+	requiredPods := make([]*kube_api.Pod, 0)
+	otherPods := make([]*kube_api.Pod, 0)
+	for _, p := range podsOnNode.Items {
+		pod := &p
+
+		creatorRef, err := ca_simulator.CreatorRefKind(pod)
+		if err != nil {
+			return []*kube_api.Pod{}, []*kube_api.Pod{}, err
+		}
+
+		if ca_simulator.IsMirrorPod(pod) || creatorRef == "DaemonSet" || isCriticalPod(pod) {
```

Comment:
Drain removes only pods which will be automatically recreated with the same state, which means being managed by higher level abstraction, without local disk, etc.

Here we only want to avoid removing pods that has a higher or equal 'priority' than the critical pod, which means manifest pods, daemon set pods and other critical addons.